### PR TITLE
[add] py-graphql-relay: new package

### PR DESCRIPTION
**Pull request by LydDeb**
Add a dependence requested by py-cylc-flow
py-cylc-flow --> py-graphene --> py-graphql-relay

### CODE REVIEW

**Review by tldahlgren — COMMENTED** [review]
FYI. Confirmed the home page and version sha256.

**Comment by tldahlgren** [comment]
Added the python package guru as a reviewer since he's much more thorough regarding dependencies, etc.

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyGraphqlRelay(PythonPackage):
+    """Relay library for graphql-core."""
+
+    homepage = "https://github.com/graphql-python/graphql-relay-py"
+    pypi = "graphql-relay/graphql-relay-3.2.0.tar.gz"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-relay/package.py]:
```suggestion
    pypi = "graphql-relay/graphql-relay-2.0.1.tar.gz"
```
Doesn't break anything, just helps with `spack url summary`

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+    maintainers("LydDeb")
+
+    version("2.0.1", sha256="870b6b5304123a38a0b215a79eace021acce5a466bf40cd39fa18cb8528afabb")
+
+    depends_on("python@3.6:3.10", type=("build", "run"))
+    depends_on("py-setuptools@59:69", type="build")
+    depends_on("py-poetry-core@1", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-relay/package.py]: I don't see these for 2.0.1

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    depends_on("py-graphql-core@2.2:2", type=("build", "run"), when="@2")
+    depends_on("py-six@1.12:", type=("build", "run"), when="@2")
+    depends_on("py-promise@2.2:2", type=("build", "run"), when="@2")
+    depends_on("py-typing-extensions@4.1:4", type=("build", "run"), when="^python@:3.7")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-relay/package.py]: I don't see this for 2.0.1

**Review by adamjstewart — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py (hunk 1 above)
Comment (LydDeb) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: It's okay with me. I first wrote this recipe for version 3.2.0 but version 2.0.1 was ultimately used. I fear that the following comments are due to leftovers from my first draft.

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py (hunk 2 above)
Comment (LydDeb) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: Absolutely, these are dependencies of version 3.2.0.
In the  pyproject.toml of the version 3.2.0 there are:
```
[build-system]
requires = ["poetry_core>=1,<2", "setuptools>=59,<70"]
```
Regarding version 2.0.1, there are only “setuptools” for the build backend but no version constraints. I'm trying to determine the maximum version of 'setuptools' for version 2.0.1

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py (hunk 3 above)
Comment (LydDeb) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: Indeed, the dependency 'py-typing-extensions' appear only in version 3.2.0. I would like to remove it.
But the others dependencies are in the 'setup.py' of the version 2.0.1:
```
    install_requires=[
        'six>=1.12',
        'graphql-core>=2.2,<3',
        'promise>=2.2,<3'
    ],
```

**Comment by tldahlgren** [comment]
@LydDeb You'll need to rebase to develop to pick up the fix for the CI failure.

**Comment by LydDeb** [comment]
Thank you for your review @tldahlgren and @adamjstewart, I modified the recipe with respect to  your remarks.

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    """Relay library for graphql-core."""
+
+    homepage = "https://github.com/graphql-python/graphql-relay-py"
+    pypi = "graphql-relay/graphql-relay-2.0.1.tar.gz"
+
+    maintainers("LydDeb")
+
+    version("2.0.1", sha256="870b6b5304123a38a0b215a79eace021acce5a466bf40cd39fa18cb8528afabb")
+
+    depends_on("python@3.4:3.8", type=("build", "run"), when="@2")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-relay/package.py]: Where do you see that newer versions of Python aren't supported? From testing?

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py (hunk 3 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: Yep, I was only commenting on typing-extensions

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py (hunk 4 above)
Comment (LydDeb) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: From commit message. Take a look at https://github.com/graphql-python/graphql-relay-py/commit/26036ef2a9da755052bec48b9449178b17652528

Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: That comment documents that 3.8 is supported but it doesn't say that 3.9 isn't. The versions documented in `classifiers` are just the ones that are tested. The package will likely work with newer versions of Python if you test them. I don't restrict Python unless I know that it won't work.

File: var/spack/repos/builtin/packages/py-graphql-relay/package.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    maintainers("LydDeb")
+
+    version("2.0.1", sha256="870b6b5304123a38a0b215a79eace021acce5a466bf40cd39fa18cb8528afabb")
+
+    depends_on("py-setuptools", type="build")
+    depends_on("py-graphql-core@2.2:2", type=("build", "run"), when="@2")
+    depends_on("py-six@1.12:", type=("build", "run"), when="@2")
+    depends_on("py-promise@2.2:2", type=("build", "run"), when="@2")
+    # spack test failed with message: ModuleNotFoundError: No module named rx
+    depends_on("py-rx", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-relay/package.py]: Tell me more about this error? I grepped the entire source directory but "rx" does not exist.

Comment (LydDeb) [reply var/spack/repos/builtin/packages/py-graphql-relay/package.py]: My bad, I did not notice that was a *ModuleNotFoundError* from py-graphql-**core** and not py-graphql-**relay**:
```
==> [2023-09-26-10:09:44.115284] test: test_imports_graphql: checking import of graphql
==> [2023-09-26-10:09:44.115514] '/opt/spack/linux-rocky8-zen/gcc-8.5.0/python-3.8.10-lesqjyytmmorjygjfy5aidyqlkh4rbgg/bin/python3.8' '-c' 'import graphql'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/opt/spack/linux-rocky8-zen/gcc-8.5.0/py-graphql-core-2.3.2-bgivakkwpd7st77auxza4yosiflf4whm/lib/python3.8/site-packages/graphql/__init__.py", line 27, in <module>
    from .graphql import graphql
  File "/opt/spack/linux-rocky8-zen/gcc-8.5.0/py-graphql-core-2.3.2-bgivakkwpd7st77auxza4yosiflf4whm/lib/python3.8/site-packages/graphql/graphql.py", line 1, in <module>
    from .execution import ExecutionResult
  File "/opt/spack/linux-rocky8-zen/gcc-8.5.0/py-graphql-core-2.3.2-bgivakkwpd7st77auxza4yosiflf4whm/lib/python3.8/site-packages/graphql/execution/__init__.py", line 21, in <module>
    from .executor import execute, subscribe
  File "/opt/spack/linux-rocky8-zen/gcc-8.5.0/py-graphql-core-2.3.2-bgivakkwpd7st77auxza4yosiflf4whm/lib/python3.8/site-packages/graphql/execution/executor.py", line 11, in <module>
    from rx import Observable
ModuleNotFoundError: No module named 'rx'
```
I will remove this dependence in the recipe of py-graphql-relay and add it in the recipe of py-graphql-core for version 2.3.2

File: var/spack/repos/builtin/packages/py-graphql-core/package.py [hunk 6]
```diff
@@ -21,3 +21,6 @@ class PyGraphqlCore(PythonPackage):
 
     depends_on("python@3.6:3", type=("build", "run"))
     depends_on("py-setuptools", type="build")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-graphql-core/package.py]:
```suggestion
    depends_on("py-poetry@1", when="@3:", type="build")
    depends_on("py-setuptools", when="@2", type="build")
```

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by LydDeb** [comment]
@adamjstewart Do I need to make any other changes?

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**